### ISSUE TITLE
edition idioms: Trickiness with $crate in local macros

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#56683]
[Originally reported with Cargo](https://github.com/rust-lang/cargo/issues/6398) this code:

```rust
#![warn(rust_2018_compatibility)]
extern crate serde;
#[macro_use]
extern crate serde_derive;

use serde::{Deserialize, Deserializer};

macro_rules! event {
    (
        pub struct $name:ident() {}
    ) => {
        #[derive(Deserialize)]
        pub struct $name {
            pub event_type: $crate::EventType,
        }
    }
}

pub enum EventType {}

event! {
    pub struct CustomEvent() {}
}

impl<'de> Deserialize<'de> for EventType {
    fn deserialize<D>(_deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        unimplemented!()
    }
}
```

[gives an erroneous suggestion](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=3db73fbef03a49165533b0a47be48c22) 

which causes it to fail to be migrated by `cargo fix`

**Comment by alexcrichton** [comment]
@ehuss also notes that https://github.com/rust-lang/rust/issues/56622 may have an impact on the severity of this

**Comment by ehuss** [comment]
This appears to be fixed now, presumably by #57155.

**Comment by alexcrichton** [comment]
Yay!
